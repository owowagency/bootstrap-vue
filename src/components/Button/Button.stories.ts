import Btn from '.';
import {sizes} from '@/composables/useSize';
import {variants} from '@/composables/useVariant';

export default {
    title: 'Components/Button',
    component: Btn,
    argTypes: {
        disabled: {
            control: {type: 'boolean'},
        },
        outline: {
            control: {type: 'boolean'},
        },
        href: {
            control: {type: 'text'},
        },
        is: {
            control: {type: 'text'},
        },
        text: {
            control: {type: 'text'},
        },
        size: {
            control: {type: 'select'},
            options: sizes,
        },
        to: {
            control: {type: 'text'},
        },
        variant: {
            control: {type: 'select'},
            options: variants,
        },
    },
};

const template = (args) => ({
    components: {Btn},
    setup() {
        return {args};
    },
    template: '<Btn v-bind="args">{{ args.text || "Default" }}</Btn>',
});

export const Default = template.bind({});

const variantsTemplate = (args) => ({
    components: {Btn},
    setup: () => ({args, variants}),
    template: '<Btn v-for="v of variants" class="me-1 mb-1" :variant="v" v-bind="args">{{ v }}</Btn>',
});

export const Variants = variantsTemplate.bind({});

export const Outline = variantsTemplate.bind({});

Outline.args = {
    outline: true,
};

export const Sizing = (args) => ({
    components: {Btn},
    setup: () => ({args, sizes}),
    template: '<Btn v-for="s of sizes" class="me-1" :size="s" v-bind="args">Button {{ s }}</Btn>',
});

export const Disabled = template.bind({});

Disabled.args = {
    disabled: true,
    text: 'Disabled',
};

export const Anchor = template.bind({});

Anchor.args = {
    href: 'https://github.com/owowagency/bootstrap-vue',
    target: '_blank',
    text: 'Anchor',
};
