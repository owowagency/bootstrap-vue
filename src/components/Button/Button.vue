<template>
    <Component
        :is="is"
        class="btn"
        :class="classes"
        :disabled="disabled"
        :href="href"
        :to="to"
    >
        <!-- @slot Displays content inside the button -->
        <slot />
    </Component>
</template>

<script lang="ts">
import {PropType, computed} from 'vue';
import useDisabled, {disabledProps} from '@/composables/useDisabled';
import useSize, {sizeProps} from '@/composables/useSize';
import useVariant, {variantProps} from '@/composables/useVariant';
import useClasses from '@/composables/useClasses';
</script>

<script lang="ts" setup>
const props = defineProps({
    ...disabledProps,
    href: {
        type: String,
        default: undefined,
    },
    is: {
        type: String,
        default: undefined,
    },
    /**
     * Changes the button to an outline variant
     */
    outline: {
        type: Boolean,
        default: false,
    },
    ...sizeProps,
    to: {
        type: [Object, String] as PropType<string|Record<string, unknown>>,
        default: undefined,
    },
    ...variantProps,
});

const variantClassTemplate = computed(() => `btn-${props.outline ? 'outline-' : ''}{0}`);

const {classes} = useClasses(computed(() => [
    useDisabled(props.disabled).disabledClass.value,
    useSize(props.size, 'btn-{0}').sizeClass.value,
    useVariant(props.variant, variantClassTemplate).variantClass.value,
]));

const is = computed(() => {
    if (props.is) {
        return props.is;
    }

    if (props.to) {
        return 'RouterLink';
    }

    if (props.href) {
        return 'a';
    }

    return 'button';
});
</script>

<docs>
```vue
<Button variant="danger">
    Delete
</Button>
```

## Imported Props

| Prop name | Description                  | Type                                             | Values | Default   |
| --------- | ---------------------------- | ------------------------------------------------ | ------ | --------- |
| disabled  | Disables the component       | boolean                                          | -      | false     |
| size      | The size of the component    | [Size](../../composables/useSize) (string)       | -      | 'md'      |
| variant   | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'primary' |
</docs>
