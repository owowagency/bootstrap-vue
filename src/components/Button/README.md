# Button

```vue
<Button variant="danger">
    Delete
</Button>
```

## Imported Props

| Prop name | Description                  | Type                                             | Values | Default   |
| --------- | ---------------------------- | ------------------------------------------------ | ------ | --------- |
| disabled  | Disables the component       | boolean                                          | -      | false     |
| size      | The size of the component    | [Size](../../composables/useSize) (string)       | -      | 'md'      |
| variant   | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'primary' |

## Props

| Prop name | Description                              | Type                                  | Values | Default   |
| --------- | ---------------------------------------- | ------------------------------------- | ------ | --------- |
| href      |                                          | string                                | -      | undefined |
| is        |                                          | string                                | -      | undefined |
| outline   | Changes the button to an outline variant | boolean                               | -      | false     |
| to        |                                          | string\|Record&lt;string, unknown&gt; | -      | undefined |

## Slots

| Name    | Description                        | Bindings |
| ------- | ---------------------------------- | -------- |
| default | Displays content inside the button |          |
