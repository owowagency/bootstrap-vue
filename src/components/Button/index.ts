import Button from './Button.vue';

export * from './Button.vue';
export default Button;
