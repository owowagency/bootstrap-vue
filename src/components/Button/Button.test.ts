import {describe, expect, it} from 'vitest';
import Button from '.';
import {shallowMount} from '@vue/test-utils';

describe('template', () => {
    componentRenderTest(Button);

    componentRenderTest(Button, {props: {href: 'https://github.com/owowagency/bootstrap-vue'}});

    componentRenderTest(Button, {props: {is: 'div'}});

    componentRenderTest(Button, {props: {to: {name: 'home'}}});

    componentSlotRenderTest(Button);

    it('adds disabled property', () => {
        const wrapper = shallowMount(Button, {
            props: {disabled: true},
        });

        expect((wrapper.element as HTMLButtonElement).disabled).toBe(true);
    });

    componentWrapperClassTest(Button, {outline: true}, 'btn-outline-primary');

    componentWrapperClassTest(Button, {size: 'lg'}, 'btn-lg');

    componentWrapperClassTest(Button, {variant: 'danger'}, 'btn-danger');
});
