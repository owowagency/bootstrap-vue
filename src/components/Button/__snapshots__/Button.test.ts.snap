// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<button class=\\"btn btn-md btn-primary\\">
  <!-- @slot Displays content inside the button -->
</button>"
`;

exports[`template > renders default 2`] = `
"<a class=\\"btn btn-md btn-primary\\" disabled=\\"false\\" href=\\"https://github.com/owowagency/bootstrap-vue\\">
  <!-- @slot Displays content inside the button -->
</a>"
`;

exports[`template > renders default 3`] = `
"<div class=\\"btn btn-md btn-primary\\" disabled=\\"false\\">
  <!-- @slot Displays content inside the button -->
</div>"
`;

exports[`template > renders default 4`] = `
"<routerlink class=\\"btn btn-md btn-primary\\" disabled=\\"false\\" to=\\"[object Object]\\">
  <!-- @slot Displays content inside the button -->
</routerlink>"
`;
